New Engine: Regulator with Bypass Relay

ghisch opened the pull request:
Hello,

This PR fixes https://github.com/XavierBerger/Solar-Router-for-ESPHome/issues/49

# Add bypass engine mode for solar router

This PR adds a new engine mode that combines the benefits of both the regular engine (smooth regulation) and the on/off engine (relay control) by automatically switching between them based on power conditions.

## Goal

Using a dimmer like a robotdyn triac allows a fine control of the load. Triac emits the most heat when they are set to 100% (when all current flows) for a sustained period of time.
This approach optimizes energy efficiency by using the relay for full power situations while maintaining smooth regulation for partial power needs.

## New files added:
- `solar_router/engine_regulator_with_bypass.yaml`: New engine implementation
- `solar_router/bypass_relay.yaml`: Bypass relay control configuration

## How it works:
1. The system operates like the regular engine, using smooth regulation (0-100%)
2. When the regulator stays at 100% for `bypass_timer_threshold` regulations and still needs more power (delta > 0):
   - The regulator is set to 0%
   - The bypass relay is activated (in this order)
3. When power demand decreases (delta < 0):
   - The bypass relay is deactivated
   - The regulator resumes from 100% + delta (in this order)
   - Returns to normal regulation mode

## Notes:
- I could not find an easy way to check since how long the regulator is opened at 100%, so I made a counter `full_power_timer` which increments of 1 every time a regulation is performed, which is roughly every seconds from what I understand.
- This PR is heavily helped by AI (cursor AI)
- **Not tested yet**

XavierBerger commented on solar_router/bypass_relay.yaml:
I think that the bypass relay itself is not necessary, we already have the mechanical relay doing the same thing.

I think that the architecture should look like this: 

![image](https://github.com/user-attachments/assets/283281c2-7f03-4d30-a0f7-68e8758bc14f)
https://github.com/XavierBerger/Solar-Router-for-ESPHome/blob/feature/engine_with_bypass/docs/images/packages.drawio.png

Variable output is plugged on SSR or Triac and ON/OFF output is plugged on Mechanical relay.

As variables regulator just care about "regulator_opening" and mechanical relay just care about "energy_divertion", just adding 2 regulators will do the job.

ghisch replied:
> I think that the bypass relay itself is not necessary, we already have the mechanical relay doing the same thing.

Agreed, will do the change soon.

> Variable output is plugged on SSR or Triac and ON/OFF output is plugged on Mechanical relay.

Maybe something for later but I think we should be able to define the relay for the ON/OFF (wether it's for the bypass or engine) as mechanical or SSR. As SSR are much faster than mechanical, using them in the Bypass would allow more smooth operations.

XavierBerger reviewed (commented):
Thanks for contribution, The algorithm looks good.
Before merging this code, it is also required to update the documentation.
I created a branch and update the Pull Request target. I'll use this branch to review the code more in detail before merging to main.

ghisch commented:
> Thanks for contribution, The algorithm looks good. Before merging this code, it is also required to update the documentation. I created a branch and update the Pull Request target. I'll use this branch to review the code more in detail before merging to main.

Ok thanks. I will work on the documentation later on. I'm not able to do it atm.

FYI:
- I'm still waiting for components delivery to be able to test this engine.
- My water heater has 3 heating resistors, I'm looking to create another engine that control the first resistor with regulator + relay and the last 2 with relays. So the triac would only have to regulate 1/3 of the water heater total power.

XavierBerger commented:
> My water heater has 3 heating resistors, I'm looking to create another engine that control the first resistor with regulator + relay and the last 2 with relays. So the triac would only have to regulate 1/3 of the water heater total power.

Nice :+1: 

This regulator is very interesting and would very benefit of the PR #38 : With this new way of setting variables for packages, it will be possible to use 1 package more than one time... Until this PR can be merged (dependency from ESPHome: https://github.com/esphome/esphome/pull/7606) we have to continue to use `substitutions` and, I think, you will have to duplicate the code of regulators...

When you are ready, you can propose a PR, and I'll be happy to discuss the implementation.

ghisch commented:
Hey 👋  I removed the `bypass_relay.yaml` and added the doc.

@XavierBerger may I ask you to create the wiring schema for the `How to wire the bypass relay` part ?

> When the regulator stays at 100% for bypass_timer_threshold regulations and still needs more power (delta > 0):
> - The regulator is set to 0%
> - The bypass relay is activated (in this order)
> 
> When power demand decreases (delta < 0):
> - The bypass relay is deactivated
> - The regulator resumes from 100% + delta (in this order)

Writing the doc I realized that it might not be smart to do it in this order, because it might creates sudden important change of consumption. Reversing the order should be smoother. This question will be answered with testing.

ghisch commented:
Using `regulator_mecanical_relay.yaml` with `regulator_triac.yaml` is not possible with current implementation because they both share `regulator_gate_pin` substitution.

We either need:
1. A dedicated yaml for the bypass relay as originaly pushed in this PR (this would duplicate code)
2. To change the name of the substitution in either of both packages (this would break compatibility for some people)

WDYT ?

XavierBerger commented:
> To change the name of the substitution in either of both packages (this would break compatibility for some people)

I think, it makes more sense to have different pin name. Even if it requires an update of the configuration for users already using the project. I will write a warning when I'll publish the next release.

Same comment for the other question. I prefer not duplicate the code.

ghisch commented:
Last commit, for all vars related to the mechanical relay, changed:
- `regulator` to `relay_regulator`
- `regulator_gate_pin` to `relay_ regulator_gate_pin`
- `regulation_control` to `relay_regulation_control`

I tested the engine, it works as intended 👍 

I faced an issue with `safety_limit` definition while testing this new engine. I was constantly having the following error despite the inclusion of `engine_common.yaml`.
```log
couldn't find id 'safety_limit'. please check you have defined an id with that name in your configuration.
```
The error was not present while using `engine.yaml`. So it's related to the new engine. My guess is that as the `full_power_timer` counter is defined as a globals like `safety_limit`, it creates circular dependency or something like that. My quick and dirty fix to still be able to perform my tests was to copy/paste `safety_limit` definition inside the file of the new engine as shown [here](https://github.com/ghisch/Solar-Router-for-ESPHome/blob/96a0998d84b7ae5db6d21bf836b8defc5ea1d7ab/solar_router/engine_regulator_with_bypass.yaml#L15-L18)

XavierBerger commented:
Did you manage to test this branch?  Can I merge it?
I would like to merge it before making too much change in the code which may generate conflicts.

ghisch commented:
Yes, I'm testing it since 2 days it does work but I have one issue: My esp sometimes reboot when the relay turn off. I think it's because of the noise generated by the triac ? I'm not sure this is way beyound my knowledge.
I'm also questioning my electrical wiring of the bypass
1. Triac Line input plugged to NC relay port (current implementation)
    - Relay switched OFF -> Only Triac is energized
    - Relay switched ON -> Only Load is energized
    - Relay switching to ON or to OFF -> Neither Triac nor Load are energized (for a small amount of time)
2. Triac Line input plugged to COM relay port
    - Relay switched OFF -> Only Triac is energized
    - Relay switched ON -> Triac & Load are energized (but Triac is shorted by relay)
    - Relay switching to ON or to OFF -> Triac is energized
    - The NC port is not used and thus floating

I'll have a test on option 2 when I can. Sorry I'll be busy in the next days.

![Sans-titre-2025-01-10-0131](https://github.com/user-attachments/assets/76e7213b-a308-4e2c-a0d6-dc9eeb01213c)

ghisch commented:
Option 2 (Triac Lin to Relay COM) is the way to go:
- https://www.st.com/resource/en/application_note/an4993-how-to-implement-a-scr-or-a-triac-in-a-hybride-relay-application-stmicroelectronics.pdf
- https://www.reddit.com/r/ElectricalEngineering/comments/dkdyh1/comment/f4f39zb/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button
- I was seeing sparks when switch ON or OFF the relay

ghisch commented:
Last commit:
- Updated the code of the engine (same output, just more efficient, cleaner, and clearer)
- Updated the comments
- Had to copy `safety_limit`inside or the code wouldn't compile (I'm suspecting circular dependencies with engine_common)
- regulator_opening is changed after relay is turned on, before relay is turned off ensuring smooth operation

ghisch commented:
Energy counter doesn't work with this engine because it's using `regulator_opening` and not `regulator_status` (but it can't use the last one because it's only defined in the regulation script). When the bypass is enabled, the energy counter says no energy is diverted because `regulator_opening` is at 0.

My proposal is:
- Rename all `regulator_status` to `router_status`, this way we keep regulator=dimmer and router=whole system.
- Define `router_status` as a number component (just like `regulator_opening`). This way we could control the router status easily from HA (if exposed to HA), have access to current value of `router_status` in script without having to recompute from the current state.
- Change all use of `regulator_opening` to `router_status` (like in the energy counter for example).

Let me know what you think of this.

XavierBerger commented:
I agree, the `router_status`, `divertion_level` or `routing_level` is not the same as the `regulator opening`, since we can have bypass.  
It is true that `theorical_energy_counter` has to take in account the "`level of energy diverted`" and not the `regulator_opening`.  
Actually, on my code, `regulator_status` was just a temporary variable. In your architecture, this variable is meaning something.  
So I think it is a good idea to rename it. I tried to find a name which is the most meaningful and I think that `level` should be part of the name as in the start of this response. I let you choose what is the best for your understanding.

XavierBerger commented:
The refacto is now merged into main. You can recreate PR.
